src/core/unit.c

```diff
@@ -517,52 +517,58 @@ void unit_add_to_dbus_queue(Unit *u) {
         u->in_dbus_queue = true;
 }
 
-void unit_submit_to_stop_when_unneeded_queue(Unit *u) {
-        assert(u);
+int unit_submit_to_stop_when_unneeded_queue(sd_event_source *s, uint64_t usec, void *userdata) {
```

Comment:
nah, please don't.

please have single handler, that will add the unit to all three queues. When the rate limit window is hit this might have effect on more than one of the queues. And adding to the queues is "free", hence better add to to many than to too few...

Reply:
in particular as we only maintain a single ratelimiter for all three queues. so if we hit the limit for one of the queues then we'll also hit it for the others, but then we won't check them again when the window is over.